File: wheel/src/main/java/dev/yong/wheel/utils/DensityUtils.java
package dev.yong.wheel.utils;


import dev.yong.wheel.AppManager;

/**
 * @author CoderYong
 */
public final class DensityUtils {

    private static float density = -1F;
    private static int widthPixels = -1;
    private static int heightPixels = -1;

    private DensityUtils() {
        throw new UnsupportedOperationException("Cannot be created");
    }

    public static float getDensity() {
        if (density <= 0F) {
            density = AppManager.getInstance().getApplication()
                    .getApplicationContext().getResources().getDisplayMetrics().density;
        }
        return density;
    }

    public static int dip2px(float dpValue) {
        return (int) (dpValue * getDensity() + 0.5F);
    }

    public static int px2dip(float pxValue) {
        return (int) (pxValue / getDensity() + 0.5F);
    }

    public static int getScreenWidth() {
        if (widthPixels <= 0) {
            widthPixels = AppManager.getInstance().getApplication()
                    .getApplicationContext().getResources().getDisplayMetrics().widthPixels;
        }
        return widthPixels;
    }


    public static int getScreenHeight() {
        if (heightPixels <= 0) {
            heightPixels = AppManager.getInstance().getApplication()
                    .getApplicationContext().getResources().getDisplayMetrics().heightPixels;
        }
        return heightPixels;
    }
}

File: wheel/src/main/java/dev/yong/wheel/utils/DensityUtils.kt
@file:Suppress("unused")

package dev.yong.wheel.utils

import android.content.res.Resources
import android.util.TypedValue

fun dp2px(dpValue: Int): Int {
    return TypedValue.applyDimension(
        TypedValue.COMPLEX_UNIT_DIP,
        dpValue.toFloat(), Resources.getSystem().displayMetrics
    ).toInt()
}

fun px2dip(pxValue: Int): Int {
    val density = Resources.getSystem().displayMetrics.density
    return (pxValue / density + 0.5f).toInt()
}

fun dp2px(dpValue: Float): Float {
    return TypedValue.applyDimension(
        TypedValue.COMPLEX_UNIT_DIP,
        dpValue, Resources.getSystem().displayMetrics
    )
}

fun px2dip(pxValue: Float): Float {
    val density = Resources.getSystem().displayMetrics.density
    return (pxValue / density + 0.5f)
}

/**
 * 获取屏幕宽度
 */
val screenWidth: Int
    get() {
        var widthPixels = -1
        if (widthPixels <= 0) {
            widthPixels = Resources.getSystem().displayMetrics.widthPixels
        }
        return widthPixels
    }

/**
 * 获取屏幕高度
 */
val screenHeight: Int
    get() {
        var heightPixels = -1
        if (heightPixels <= 0) {
            heightPixels = Resources.getSystem().displayMetrics.heightPixels
        }
        return heightPixels
    }